
---

# Generic form of data sets

Routinely gathered in ecology/microbiology/genomics

### Data tables

  - .important[Abundances]: read counts of species/transcripts $j$ in sample $i$
  - .important[Covariates]: value of environmental variable $k$ in sample $i$
  - .important[Offsets]: sampling effort for species/transcripts $j$ in sample $i$

### Need frameworks to model _dependencies between counts_

  - understand **environmental effects** <br/>
      $\rightsquigarrow$ explanatory models (multivariate regression, classification)
  - exhibit **patterns of diversity** <br/>
      $\rightsquigarrow$ summarize the information (clustering, dimension reduction)
  - understand **between-species interactions** <br />
      $\rightsquigarrow$ 'network' inference (variable/covariance selection)
  - correct for technical and **confounding effects** <br/>
      $\rightsquigarrow$ account for covariables and sampling effort

---

# Illustration in genomics

## scRNA data set 

The dataset `scRNA` contains the counts of the 500 most varying transcripts in the mixtures of 5 cell lines in human liver (obtained with standard 10x scRNAseq Chromium protocol).

We subsample 500 random cells and the keep the 200 most varying genes

```{r init-data, echo = FALSE}
library(PLNmodels); library(ZIPLN)
```

```{r scRNA-dataset}
data(scRNA); set.seed(1234)
train_set <- sample.int(nrow(scRNA), 500)
test_set  <- setdiff(1:nrow(scRNA), train_set)
scRNA_train <- scRNA[train_set, ]
scRNA_test  <- scRNA[test_set, ]
scRNA_train$counts <- scRNA_train$counts[, 1:200]
scRNA_test$counts  <- scRNA_test$counts[, 1:200]
```

### Covariates

- `cell_line`: the cell line of the current row (among 5)
- `total_counts`: Total number reads for that cell

---

# Table of Counts

.pull-left[
Matrix view (log-transform + total-counts normalization)

```{r glance counts, fig.height=6, fig.cap="log-scaled counts", echo = FALSE}
log_data <- sweep(log(1 + scRNA_train$counts), 1, log(scRNA_train$total_counts), "-")
log_data %>% 
  corrplot::corrplot(is.corr = FALSE,
    addgrid.col = NA,  tl.cex = .5,  cl.pos = "n", tl.pos = "n")
```
]

.pull-right[
Histogram (log-transform + total-counts normalization)
```{r glimpse Abundance, echo = FALSE, message=FALSE}
ggplot(data.frame(counts = c(log_data)), aes(counts))  + 
  geom_histogram() + theme_bw()
```
]

---

# Models for multivariate count data

### If we were in a Gaussian world...

The .important[general linear model] `r Citep(myBib, "MKB79")` would be appropriate! For each sample $i = 1,\dots,n$, 

$$\underbrace{\mathbf{Y}_i}_{\text{abundances}} =  \underbrace{\mathbf{x}_i^\top \mathbf{B}}_{\text{covariates}} + \underbrace{\mathbf{o}_i}_{\text{sampling effort}} + \boldsymbol\varepsilon_i, \quad \boldsymbol\varepsilon_i \sim \mathcal{N}(\mathbf{0}_p, \underbrace{\boldsymbol\Sigma}_{\text{between-species dependencies}})$$

null covariance $\Leftrightarrow$ independence $\rightsquigarrow$ uncorrelated species/transcripts do not interact

.content-box-red[This model gives birth to Principal Component Analysis,
 Discriminant Analysis, Gaussian Graphical Models, Gaussian Mixture models and many others $\dots$]

### With count data...

There is no generic model for multivariate counts
  - Data transformation (log, $\sqrt{}$) : quick and dirty
  - Non-Gaussian multivariate distributions `r Citep(myBib, "IYA16")`: do not scale to data dimension yet
  - .important[Latent variable models]: interaction occur in a latent (unobserved) layer

---

# The Poisson Lognormal model (PLN)

The PLN model `r Citep(myBib, "AiH89")` is a .important[multivariate generalized linear model], where 

- the counts $\mathbf{Y}_i$ are the response variables
- the main effect is due to a linear combination of the covariates $\mathbf{x}_i$
- a vector of offsets $\mathbf{o}_i$ can be specified for each sample.

.content-box-red[
$$
\mathbf{Y}_i | \mathbf{Z}_i \sim \mathcal{P}\left(\exp\{\mathbf{Z}_i\}\right), \qquad \mathbf{Z}_i \sim \mathcal{N}({\mathbf{o}_i + \mathbf{x}_i^\top\mathbf{B}},\boldsymbol\Sigma), \\
$$
]
.pull-left[The unkwown parameters are 
- $\mathbf{B}$, the regression parameters
- $\boldsymbol\Sigma$, the variance-covariance matrix
]

.pull-right[
Stacking all individuals together, 
  - $\mathbf{Y}$ is the $n\times p$ matrix of counts  
  - $\mathbf{X}$ is the $n\times d$ matrix of design
  - $\mathbf{O}$ is the $n\times p$ matrix of offsets
]

### Properties: .small[.important[over-dispersion, arbitrary-signed covariances]]

- mean: $\mathbb{E}(Y_{ij}) =  \exp \left( o_{ij} + \mathbf{x}_i^\top {\mathbf{B}}_{\cdot j} + \sigma_{jj}/2\right) >  0$
- variance: $\mathbb{V}(Y_{ij}) = \mathbb{E}(Y_{ij}) + \mathbb{E}(Y_{ij})^2 \left( e^{\sigma_{jj}} - 1 \right) > \mathbb{E}(Y_{ij})$
- covariance: $\mathrm{Cov}(Y_{ij}, Y_{ik}) = \mathbb{E}(Y_{ij}) \mathbb{E}(Y_{ik}) \left( e^{\sigma_{jk}} - 1 \right).$


