
---

# Generic form of data sets
  
Routinely gathered in ecology/microbiology/genomics 

### Data tables

  - .important[Abundances]: read counts of species/transcripts $j$ in sample $i$
  - .important[Covariates]: value of environmental variable $k$ in sample $i$
  - .important[Offsets]: sampling effort for species/transcripts $j$ in sample $i$

### Need a framework to model _dependencies between counts_ 

  - understand .important[environmental effects] <br/>
      $\rightsquigarrow$ explanatory models (multivariate regression, classification)
  - exhibit .important[patterns of diversity] <br/>
      $\rightsquigarrow$ summarize the information (clustering, dimension reduction)
  - understand .important[between-species interactions] <br />
      $\rightsquigarrow$ 'network' inference (variable/covariance selection)
  - correct for technical and .important[confounding effects] <br/>
      $\rightsquigarrow$ account for covariables and sampling effort

---

# Oaks powdery mildew data set overview

- .important[Microbial communities] sampled on the surface of $n = 116$ oak leaves
- Communities sequenced and cleaned resulting in $p=114$ OTUs (66 bacteria, 48 fungi).
- Study .important[effects of the pathogen] _E.Aphilto√Ødes_ wrt communities

The `oaks` variable consists in a special data frame ready to play with, typical from ecological data sets (try `?oaks` to get additional details).

<small>
```{r load oaks}
data("oaks")
str(oaks, max.level = 1)
```
</small>

---

# Covariates and offsets

Characterize the samples and the sampling, most important being

- `tree`: Tree status with respect to the pathogen (susceptible, intermediate or resistant)
- `distTOground`: Distance of the sampled leaf to the base of the ground
- `orientation`: Orientation of the branch (South-West SW or North-East NE)
- `readsTOTfun`: Total number of ITS1 reads for that leaf
- `readsTOTbac`: Total number of 16S reads for that leaf

```{r covariates summary}
summary(oaks$tree)
summary(oaks$distTOground)
```

$\rightsquigarrow$ `readsTOTfun` and `readsTOTbac` are candidate for modeling sampling effort as offsets

---

# Abundance table (I)

```{r glimpse Abundance}
oaks$Abundance %>% as_tibble() %>% 
  dplyr::select(1:10) %>% 
  head() %>% knitr::kable(format = "html")
```

---

# Abundance table (II)

```{r glance Abundances, fig.height=6}
log(1 + oaks$Abundance) %>% 
  corrplot::corrplot(is.corr = FALSE,
    addgrid.col = NA,  tl.cex = .5,  cl.pos = "n")
```

