```{r setup-oaks, include=FALSE, warning=FALSE, message=FALSE}
library(tidyverse)  # data manipulation
library(PLNmodels)  # plot of covariance/correlation matrices
library(corrplot)   # plot of covariance/correlation matrices
library(ggraph)
library(igraph)
library(RColorBrewer)

fancy_network <- function(net, taxonomy, edge_color = c("#F8766D", "#00BFC4")) {
  
  mygraph <- plot(net, plot = FALSE)

  V(mygraph)$type <- as.character(taxonomy)
  E(mygraph)$color   <- factor(ifelse(E(mygraph)$weight > 0, edge_color[1], edge_color[2]), 
                               levels = edge_color)
  
  angle <- 90 - 360 * 1:igraph::gorder(mygraph) / igraph::gorder(mygraph)
  V(mygraph)$hjust <- ifelse(angle < -90, 1, 0)
  V(mygraph)$angle <- ifelse(angle < -90, angle + 180, angle)
  
  p <- ggraph(mygraph, layout = 'linear', circular = TRUE) + 
    geom_edge_arc(aes(edge_colour = color, edge_width = abs(weight)), alpha = 0.5, show.legend = FALSE) +
    geom_node_text(aes(label = type, x = x*1.1, y=y*1.1, angle = angle, hjust = hjust), size = 2, alpha = 1) +
    geom_node_point(aes(colour = type, size = size, alpha = 0.2)) +
    scale_colour_manual(values = c(brewer.pal(6,"Paired"),"maroon4","springgreen", "tan4","plum","brown4","burlywood",brewer.pal(8, "Dark2"), "navy")) +
    #    scale_colour_manual(values = c(brewer.pal(12,"Paired"),brewer.pal(8, "Dark2"), "navy")) +
    scale_size_continuous( range = c(0.1,10) ) +
    theme_void() +
    theme(plot.margin=unit(c(0,0,0,0),"cm"),) + 
    guides(alpha = "none", size = "none") + 
    expand_limits(x = c(-1.5, 1.5), y = c(-1.5, 1.5))
  
  invisible(p)
}
```

```{r future, echo = FALSE}
options(future.fork.enable = TRUE)
future::plan("multicore", workers = 10)
```

---

# Illustration on ecological data (eDNA)

### Oaks powdery mildew data set

`r Citet(myBib, "JAL16")` Study .important[effects of the pathogen] _E.AphiltoÃ¯des_ (mildew) wrt bacterial and microbial communities

### Species Abundances

- .important[Microbial communities] sampled on the surface of $n = 116$ oak leaves
- Communities sequenced and cleaned resulting in $p=114$ OTUs (66 bacteria, 48 fungi).

### Covariates and offsets

Characterize the samples and the sampling, most important being

- `tree`: Tree status with respect to the pathogen (susceptible, intermediate or resistant)
- `distTOground`: Distance of the sampled leaf to the base of the ground
- `orientation`: Orientation of the branch (South-West SW or North-East NE)
- `readsTOTfun`: Total number of ITS1 reads for that leaf
- `readsTOTbac`: Total number of 16S reads for that leaf

---

# Abundance table

.pull-left[
Data table
```{r glimpse Abundance, echo = FALSE}
data(oaks)
oaks$Abundance %>% as_tibble() %>% 
  dplyr::select(5:10) %>% 
  rmarkdown::paged_table()
```
]
.pull-right[
Matrix of count (log-scale)

```{r glance Abundances, fig.height=6, fig.cap="log-scaled counts", echo = FALSE}
log(1 + oaks$Abundance) %>% 
  corrplot::corrplot(is.corr = FALSE,
    addgrid.col = NA,  tl.cex = .5,  cl.pos = "n")
```
]

---
#  PLN with offsets and covariates (1)

## Offset: .small[modeling sampling effort]

The predefined offset uses the total sum of reads, accounting for technologies specific to fungi and bacteria:

```{r simple PLN offsets, cache = TRUE, results = FALSE}
M01_oaks <- PLN(Abundance ~ 1 + offset(log(Offset)) , oaks)
```

## Covariates: .small[tree and orientation effects ('ANOVA'-like) ]

The `tree` status is a natural candidate for explaining a part of the variance.

- We chose to describe the tree effect in the regression coefficient (mean)
- A possibly spurious effect regarding the interactions  between species (covariance).

```{r PLN covariate oaks, cache = TRUE, results = FALSE}
M11_oaks <- PLN(Abundance ~ 0 + tree + offset(log(Offset)), oaks)
```

What about adding more covariates in the model, e.g. the orientation?

```{r PLN regroup oaks modalities, cache = TRUE, results = FALSE}
M21_oaks <- PLN(Abundance ~  0 + tree + orientation + offset(log(Offset)), oaks)
```

---
#  PLN with offsets and covariates (2)

There is a clear gain in introducing the tree covariate in the model:

```{r PLN covariate oaks results}
rbind(M01 = M01_oaks$criteria,
      M11 = M11_oaks$criteria, M21 = M21_oaks$criteria) %>% 
  knitr::kable(format = "html")
```

Looking at the coefficients $\mathbf{B}$ associated with `tree` bring additional insights:

```{r oaks matrix plot, fig.width=14, fig.height=2, echo = FALSE}
coef(M11_oaks) %>% t() %>% corrplot(method = "color", is.corr = FALSE, tl.cex = 1, cl.pos = "n")
```

---

# Discriminant Analysis

Use the `tree` variable for grouping (`grouping` is a factor of group to be considered)

```{r PLNLDA oaks, results = FALSE}
myLDA_tree <- 
  PLNLDA(Abundance ~ 1 + offset(log(Offset)), grouping = oaks$tree, data = oaks)
```

.pull-left[
```{r plot oaks1, echo = FALSE, fig.height = 6}
plot(myLDA_tree, map = "individual")
```
]

.pull-right[
```{r plot oaks2, echo = FALSE, fig.height = 6}
plot(myLDA_tree, map = "variable")
```
]

---
# A PCA analysis of the oaks data set

```{r PLNPCA offset, cache.rebuild= TRUE, warning=FALSE, message=FALSE, results='hide'}
PCA_offset <- PLNPCA(Abundance ~ 1 + offset(log(Offset)), data = oaks, ranks = 1:30)
```

<small>
```{r PCA offset vizu tree, fig.width=6, fig.height=6, fig.align="center", echo=FALSE}
getBestModel(PCA_offset, "BIC") %>% factoextra::fviz_pca_biplot(
  select.var = list(contrib = 10), addEllipses = TRUE, habillage = oaks$tree,
  title = "Biplot (10 most contributing species)"
  ) + labs(col = "tree status") + scale_color_viridis_d()
```
</small>

---
# PCA: removing covariate effects

To hopefully find some hidden effects in the data, we can try to remove confounding ones:

```{r PCA covariate tree, cache.rebuild= TRUE, warning=FALSE, message=FALSE, results='hide'}
PCA_tree <- 
  PLNPCA(Abundance ~ 0 + tree + offset(log(Offset)), data = oaks, ranks = 1:30)
```

```{r PCA covariate tree plot, echo = FALSE, fig.align="center", fig.width=6, fig.height=6}
PCA_tree %>% getBestModel("BIC") %>% 
factoextra::fviz_pca_biplot(
  select.var = list(contrib = 10), col.ind = oaks$distTOground,
  title = "Biplot (10 most contributing species)"
  ) + labs(col = "distance (cm)") + scale_color_viridis_c()
```

---
# Clustering of the oaks samples

```{r PLNmixture-oaks, cache.rebuild = TRUE, results='hide'}
PLN_mixtures <-
   PLNmixture(Abundance ~ 1 + offset(log(Offset)), data = oaks, clusters = 1:3)
myPLN_mix <- getModel(PLN_mixtures, 3)
```

.pull-left[
```{r PLNclustering1fake, eval = FALSE}
myPLN_mix$plot_clustering_pca()
```

```{r PLNclustering1, echo = FALSE, fig.align='center', fig.height=6}
myPLN_mix$plot_clustering_pca(main = 'clustering memberships in individual factor map')
```
]

.pull-right[
```{r PLNclustering2, fig.align='center', fig.height=6}
myPLN_mix$plot_clustering_data()
```
]

---
# Network inference

```{r PLNnetwork, cache.rebuild= TRUE, warning=FALSE, message=FALSE, results='hide'}
networks <- PLNnetwork(Abundance ~ 0 + tree + offset(log(Offset)), data = oaks)
```

.pull-left[
```{r PLNnetwork criteria, echo = FALSE, fig.align='center', fig.height=6, warning=FALSE}
plot(networks)
```
]

.pull-right[
```{r PLNnetwork network, echo = FALSE, fig.align='center', fig.height=6, warning=FALSE}
type <- rep(c("bacterial", "fungal", "Alphiltoides"), c(66,114-67,1))
net <- fancy_network(getBestModel(networks, "EBIC"), type)
print(net)
```
]


```{r future off, echo=FALSE}
future::plan("sequential")
```

---

# Availability

### Help and documentation

- github group <https://github.com/pln-team>
- PLNmodels website <https://pln-team.github.io/PLNmodels> 

### R/C++ Package `PLNmodels`

Last stable release on CRAN,  development version available on GitHub).

```{r install, eval=FALSE, tidy=FALSE}
install.packages("PLNmodels")
remotes::install_github("PLN-team/PLNmodels@dev")
```

```{r loading package, message=FALSE}
library(PLNmodels)
packageVersion("PLNmodels")
```

### Python module `pyPLNmodels`

A Python/PyTorch implementation is about to be published

---
# Simple torch example in `R` 

```{r, cache.rebuild=TRUE,  torch-vs-nlopt}
data("oaks")
system.time(myPLN_torch <-
              PLN(Abundance ~ 1  + offset(log(Offset)),
                  data = oaks, control = list(backend = "torch", trace = 0)))
system.time(myPLN_nlopt <-
              PLN(Abundance ~ 1  + offset(log(Offset)),
                  data = oaks, control = list(backend = "nlopt", trace = 0)))
```

```{r}
myPLN_torch$loglik
myPLN_nlopt$loglik
```

