<!DOCTYPE html>
<html lang="fr"><head>
<script src="slides_files/libs/clipboard/clipboard.min.js"></script>
<script src="slides_files/libs/quarto-html/tabby.min.js"></script>
<script src="slides_files/libs/quarto-html/popper.min.js"></script>
<script src="slides_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="slides_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="slides_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="slides_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="slides_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.353">

  <meta name="author" content="Bastien BatardiÃ¨re">
  <meta name="author" content="Joon Kwon">
  <meta name="author" content="Mahendra Mariadassou">
  <meta name="author" content="Julien Chiquet">
  <title>pyPLNmodels: analyze multivariate count data</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="slides_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="slides_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="slides_files/libs/revealjs/dist/theme/quarto.css">
  <link href="slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="slides_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="slides_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
  <script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="_extensions/davidcarayon/inrae/ressources/assets/sigle-inrae.png" data-background-opacity="0.5" data-background-position="left" data-background-size="40%" data-footer="" class="quarto-title-block center">
  <h1 class="title"><code>pyPLNmodels</code>: analyze multivariate count data</h1>
  <p class="subtitle">Groupe de travail StatOmique</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
<strong>Bastien BatardiÃ¨re</strong> 
</div>
        <p class="quarto-title-affiliation">
            MIA Paris-Saclay, INRAE
          </p>
    </div>
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Joon Kwon 
</div>
        <p class="quarto-title-affiliation">
            MIA Paris-Saclay, INRAE
          </p>
    </div>
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Mahendra Mariadassou 
</div>
        <p class="quarto-title-affiliation">
            MaIAGE, INRAE
          </p>
    </div>
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Julien Chiquet 
</div>
        <p class="quarto-title-affiliation">
            MIA Paris-Saclay, INRAE
          </p>
    </div>
</div>

  <p class="date">11 mars 2025</p>
</section>
<section id="overview" class="slide level2">
<h2>Overview</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<ul>
<li><strong>Data</strong>: single cell RNA-seq data</li>
<li><strong>Introduction</strong>: PLN models</li>
<li><strong>Run</strong> some examples of <span class="math inline">\(\texttt{pyPLNmodels}\)</span></li>
<li>Focus on <strong>ZIPln</strong> and <strong>PlnAR</strong></li>
</ul>
</section>
<section class="slide level2">

</section>
<section>
<section id="motivation-multivariate-count-data" class="title-slide slide level1 center">
<h1>Motivation: multivariate count data</h1>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
</section>
<section id="single-cell-rna-seq-data." class="slide level2">
<h2>Single cell RNA-seq data.</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<p><img src="figures/matrix.png" width="100%"></p>
</section>
<section id="data-tables" class="slide level2">
<h2>Data tables</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<ul>
<li><span class="math inline">\(\texttt{endog}\)</span>: <span class="math inline">\(\mathbf Y\)</span>, endogenous count variables (read counts of transcripts <span class="math inline">\(j\)</span> in sample <span class="math inline">\(i\)</span>)</li>
<li><span class="math inline">\(\texttt{exog}\)</span>: <span class="math inline">\(\mathbf X\)</span>, exogenous variables (available covariates)</li>
<li><span class="math inline">\(\texttt{offsets}\)</span>: <span class="math inline">\(\mathbf O\)</span>, sampling effort for transcripts <span class="math inline">\(j\)</span> in sample <span class="math inline">\(i\)</span></li>
</ul>
</section>
<section id="goals" class="slide level2">
<h2>Goals</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<ul>
<li>understand covariates (regression, classification)</li>
<li>exhibit patterns (clustering, dimension reduction)</li>
<li>understand between-genes interactions (covariance analysis)</li>
<li>explain zero inflation</li>
</ul>
</section>
<section class="slide level2">

<h4 id="illustration-scmark-dataset-scmark">Illustration: scMARK dataset <span class="citation" data-cites="scMARK">(<a href="#/references" role="doc-biblioref" onclick="">Swechha et al. 2021</a>)</span></h4>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1"></a><span class="im">from</span> pyPLNmodels <span class="im">import</span> load_scrna</span>
<span id="cb1-2"><a href="#cb1-2"></a>rna <span class="op">=</span> load_scrna(n_samples <span class="op">=</span> <span class="dv">2000</span>, dim <span class="op">=</span> <span class="dv">3000</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a>rna.keys()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Returning scRNA dataset of size (2000, 3000)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="1">
<pre><code>dict_keys(['endog', 'labels', 'labels_1hot'])</code></pre>
</div>
</div>
</section>
<section id="data-description" class="slide level2">
<h2>Data description</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="cell" data-execution_count="3">
<div class="cell-output cell-output-display" data-execution_count="2">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>

<table class="dataframe" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header" style="text-align: right;">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">FTL</th>
<th data-quarto-table-cell-role="th">MALAT1</th>
<th data-quarto-table-cell-role="th">FTH1</th>
<th data-quarto-table-cell-role="th">TMSB4X</th>
<th data-quarto-table-cell-role="th">CD74</th>
<th data-quarto-table-cell-role="th">SPP1</th>
<th data-quarto-table-cell-role="th">APOE</th>
<th data-quarto-table-cell-role="th">B2M</th>
<th data-quarto-table-cell-role="th">ACTB</th>
<th data-quarto-table-cell-role="th">HLA-DRA</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">RBM22</th>
<th data-quarto-table-cell-role="th">FBXL15</th>
<th data-quarto-table-cell-role="th">SNAPIN</th>
<th data-quarto-table-cell-role="th">OSCAR</th>
<th data-quarto-table-cell-role="th">NUDT22</th>
<th data-quarto-table-cell-role="th">CASP8</th>
<th data-quarto-table-cell-role="th">DNAJA4</th>
<th data-quarto-table-cell-role="th">COA6</th>
<th data-quarto-table-cell-role="th">SMG1</th>
<th data-quarto-table-cell-role="th">PLEKHF1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">count</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>...</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
<td>2000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">mean</td>
<td>62</td>
<td>119</td>
<td>36</td>
<td>65</td>
<td>34</td>
<td>10</td>
<td>13</td>
<td>55</td>
<td>33</td>
<td>22</td>
<td>...</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">std</td>
<td>224</td>
<td>112</td>
<td>95</td>
<td>92</td>
<td>88</td>
<td>69</td>
<td>67</td>
<td>62</td>
<td>61</td>
<td>60</td>
<td>...</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">min</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>...</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">25%</td>
<td>2</td>
<td>35</td>
<td>2</td>
<td>15</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>20</td>
<td>5</td>
<td>0</td>
<td>...</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">50%</td>
<td>5</td>
<td>94</td>
<td>7</td>
<td>36</td>
<td>3</td>
<td>0</td>
<td>0</td>
<td>41</td>
<td>14</td>
<td>1</td>
<td>...</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">75%</td>
<td>17</td>
<td>169</td>
<td>22</td>
<td>74</td>
<td>16</td>
<td>0</td>
<td>1</td>
<td>66</td>
<td>36</td>
<td>7</td>
<td>...</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">max</td>
<td>4129</td>
<td>1171</td>
<td>2006</td>
<td>1179</td>
<td>1010</td>
<td>1792</td>
<td>1197</td>
<td>771</td>
<td>1463</td>
<td>500</td>
<td>...</td>
<td>4</td>
<td>4</td>
<td>4</td>
<td>5</td>
<td>5</td>
<td>3</td>
<td>9</td>
<td>7</td>
<td>5</td>
<td>14</td>
</tr>
</tbody>
</table>

<p>8 rows Ã 3000 columns</p>
</div>
</div>
</div>
</section>
<section id="data-description-1" class="slide level2">
<h2>Data description</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="cell" data-execution_count="4">
<div class="cell-output cell-output-stdout">
<pre><code>T_cells_CD4+    734
T_cells_CD8+    720
Macrophages     546
Name: standard_true_celltype_v5, dtype: int64</code></pre>
</div>
</div>
</section>
<section id="data-description-2" class="slide level2">
<h2>Data description</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="cell" data-execution_count="5">
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-5-output-1.png" width="1188" height="671"></p>
</div>
</div>
</section>
<section id="data-description-3" class="slide level2">
<h2>Data description</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="cell" data-execution_count="6">
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-6-output-1.png" width="1188" height="651"></p>
</div>
</div>
</section></section>
<section>
<section id="background-on-the-poisson-lognormal-family-model-inference" class="title-slide slide level1 center">
<h1>Background on <br> the Poisson-lognormal Family <br> Model, inference</h1>
<div class="footer">

</div>
</section>
<section class="slide level2">

<div class="footer">

</div>
<div class="fragment">
<!-- -->
<h3 id="the-poisson-log-normal-model">The Poisson Log Normal model</h3>
<p>PLN <span class="citation" data-cites="AiH89">(<a href="#/references" role="doc-biblioref" onclick="">Aitchison et Ho 1989</a>)</span> is a <span class="alert">multivariate generalized linear model</span>, where</p>
<ul>
<li>the counts <span class="math inline">\(\mathbf{Y}_i\in\mathbb{N}^p\)</span> are the response variables</li>
<li>the main effect is due to a linear combination of the covariates <span class="math inline">\(\mathbf{x}_i\in\mathbb{R}^d\)</span></li>
<li>a vector of offsets <span class="math inline">\(\mathbf{o}_i\in\mathbb{R}^p\)</span> can be specified</li>
</ul>
</div>
<div class="fragment">
<p><span class="math display">\[
\quad \mathbf{Y}_i | \mathbf{Z}_i \sim^{\text{iid}} \mathcal{P}\left(\exp\{\mathbf{Z}_i\}\right), \qquad \mathbf{Z}_i \sim \mathcal{N}({\underbrace{\mathbf{o}_i + \mathbf{x}_i^\top\mathbf{B}}}_{{\boldsymbol\mu}_i},\boldsymbol\Sigma).
\]</span></p>
<p>Typically, <span class="math inline">\(n\approx 10000s\)</span>, <span class="math inline">\(p\approx 10000s\)</span>, <span class="math inline">\(d \leq 10\)</span></p>
<p><span class="math display">\[\mathbb V[Y_{ij}] &gt; \mathbb E[Y_{ij}] \]</span></p>
</div>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="why-use-pln-models" class="title-slide slide level1 center">
<h1>Why use PLN models?</h1>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-1-1">LDA</a></li><li><a href="#tabset-1-2">Classif on <span class="math inline">\(\mathbf Y\)</span></a></li><li><a href="#tabset-1-3">Classif on <span class="math inline">\(\mathbb E \left[\mathbf Z|\mathbf Y\right]\)</span></a></li></ul>
<div class="tab-content">
<div id="tabset-1-1">
<p>Suppose we aim to predict the cell type based on the scMARK dataset.</p>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1"></a><span class="im">from</span> sklearn.model_selection <span class="im">import</span> train_test_split</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="im">from</span> sklearn.discriminant_analysis <span class="im">import</span> LinearDiscriminantAnalysis <span class="im">as</span> LDA</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="kw">def</span> get_classif_error(data, y):</span>
<span id="cb5-5"><a href="#cb5-5"></a>    data_train, data_test, y_train, y_test <span class="op">=</span> train_test_split(data, y, test_size<span class="op">=</span><span class="fl">0.33</span>, random_state<span class="op">=</span><span class="dv">42</span>)</span>
<span id="cb5-6"><a href="#cb5-6"></a>    lda <span class="op">=</span> LDA()</span>
<span id="cb5-7"><a href="#cb5-7"></a>    lda.fit(data_train, y_train)</span>
<span id="cb5-8"><a href="#cb5-8"></a>    y_pred <span class="op">=</span> lda.predict(data_test)</span>
<span id="cb5-9"><a href="#cb5-9"></a>    <span class="cf">return</span> np.mean(y_pred <span class="op">!=</span> y_test)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-1-2">
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1"></a>rna <span class="op">=</span> load_scrna(n_samples <span class="op">=</span> <span class="dv">1000</span>)</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="bu">print</span>(<span class="st">'Classif error:'</span>,get_classif_error(rna[<span class="st">"endog"</span>], rna[<span class="st">"labels"</span>]))</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Returning scRNA dataset of size (1000, 100)
Classif error: 0.31212121212121213</code></pre>
</div>
</div>
</div>
<div id="tabset-1-3">
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1"></a><span class="im">from</span> pyPLNmodels <span class="im">import</span> Pln</span>
<span id="cb8-2"><a href="#cb8-2"></a>latent_variables <span class="op">=</span> Pln(rna[<span class="st">"endog"</span>]).fit().latent_variables</span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="bu">print</span>(<span class="st">"Classif error: "</span>, get_classif_error(latent_variables, rna[<span class="st">"labels"</span>]))</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Setting the offsets to zero.
Fitting a Pln model with full covariance.
Intializing parameters ...
Initialization finished.
Maximum number of iterations (400)  reached in 3.0 seconds.
Last  criterion = 5.4e-06 . Required tolerance = 1e-06
Classif error:  0.17575757575757575</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section class="slide level2">

<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<h3 id="qualitative-assessment">Qualitative assessment</h3>
<p><img src="figures/plnpca_vs_pca_french.png" width="100%"></p>
</section>
<section id="natural-extensions-of-pln" class="slide level2">
<h2>Natural extensions of PLN</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<h4 id="various-tasks-of-multivariate-analysis">Various tasks of multivariate analysis</h4>
<ul>
<li><p><strong>Dimension Reduction</strong>: rank constraint matrix <span class="math inline">\(\boldsymbol\Sigma\)</span>. <small><span class="citation" data-cites="PLNPCA">(<a href="#/references" role="doc-biblioref" onclick="">Chiquet, Mariadassou, et Robin 2018</a>)</span></small> <span class="math display">\[\mathbf{Z}_i \sim \mathcal{N}({\boldsymbol\mu}_i, \boldsymbol\Sigma = \mathbf{C}\mathbf{C}^\top), \quad \mathbf{C} \in \mathcal{M}_{pk} \text{ with orthogonal columns}.\]</span></p></li>
<li><p><strong>Network inference</strong>: sparsity constraint on inverse covariance. <small><span class="citation" data-cites="PLNnetwork">(<a href="#/references" role="doc-biblioref" onclick="">Chiquet, Mariadassou, et Robin 2019</a>)</span></small> <span class="math display">\[\mathbf{Z}_i \sim \mathcal{N}({\boldsymbol\mu}_i, \boldsymbol\Sigma = \boldsymbol\Omega^{-1}), \quad \|\boldsymbol\Omega \|_1 &lt; c.\]</span></p></li>
</ul>
</section>
<section class="slide level2">

<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<h4 id="various-tasks-of-multivariate-analysis-classification">Various tasks of multivariate analysis: Classification</h4>
<ul>
<li><p><strong>Unsupervised</strong>: mixture model in the latent space <small><span class="citation" data-cites="PLNmodels">(<a href="#/references" role="doc-biblioref" onclick="">Chiquet, Mariadassou, et Robin 2021</a>)</span></small> <span class="math display">\[\mathbf{Z}_i \mid c_i = k  \sim \mathcal{N}(\boldsymbol\mu_k, \boldsymbol\Sigma_k), \quad \text{for unknown memberships } c_i.\]</span></p></li>
<li><p><strong>Supervised</strong>: maximize separation between groups with means <small><span class="citation" data-cites="PLNmodels">(<a href="#/references" role="doc-biblioref" onclick="">Chiquet, Mariadassou, et Robin 2021</a>)</span></small> <span class="math display">\[\mathbf{Z}_i \sim \mathcal{N}(\sum_k {\boldsymbol\mu}_k \mathbf{1}_{\{c_i= k\}}, \boldsymbol\Sigma), \quad \text{for known memberships } c_i.\]</span></p></li>
</ul>
</section>
<section class="slide level2">

<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<h4 id="various-tasks-of-multivariate-analysis-zero-inflation">Various tasks of multivariate analysis: Zero-inflation</h4>
<ul>
<li><p><strong>Zero-inflation</strong>: Add a zero inflation latent variable <small><span class="citation" data-cites="PLNmodels">(<a href="#/references" role="doc-biblioref" onclick="">Chiquet, Mariadassou, et Robin 2021</a>)</span></small> <span class="math display">\[\mathbf{W}_i \sim \mathcal{B}(\sigma(\boldsymbol{\mu}^{0}_i)), \quad  \mathbf Y_i \sim (1 - \mathbf W_i)\odot \mathcal P(\exp(\mathbf Z_i))\]</span></p></li>
<li><p><strong>Dimension reduction</strong> and <strong>Zero Inflation</strong>: Add a rank constraint on <span class="math inline">\(\boldsymbol \Sigma\)</span></p></li>
</ul>
</section>
<section class="slide level2">

<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<h4 id="various-tasks-of-multivariate-analysis-autogressive-models">Various tasks of multivariate analysis: autogressive models</h4>
<ul>
<li><strong>Autoregressive</strong>: Assume dependency between individuals <span class="math display">\[\mathbf{Z}_{i+1} = \Phi \mathbf Z_i  + \epsilon_i \]</span></li>
</ul>
</section></section>
<section>
<section id="estimation" class="title-slide slide level1 center">
<h1>Estimation</h1>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
</section>
<section id="estimation-1" class="slide level2">
<h2>Estimation</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<p>Estimate <span class="math inline">\(\theta = (\mathbf{B}, {\boldsymbol\Sigma})\)</span>, predict the <span class="math inline">\(\mathbf{Z}_i\)</span>, while the model marginal likelihood is</p>
<p><span class="math display">\[\begin{equation*}
p_\theta(\mathbf{Y}_i) = \int_{\mathbb{R}_p} \prod_{j=1}^p p_\theta(Y_{ij} | Z_{ij}) \, p_\theta(\mathbf{Z}_i) \mathrm{d}\mathbf{Z}_i
\end{equation*}\]</span></p>
<ul>
<li><p>Untractable likelihood</p></li>
<li><p>EM requires to evaluate (some moments of) <span class="math inline">\(p_\theta(\mathbf{Z} \,|\, \mathbf{Y})\)</span>, but there is no close form!</p></li>
<li><p><span class="math inline">\(\implies\)</span> variational EM</p></li>
</ul>
</section>
<section id="variational-inference" class="slide level2">
<h2>Variational Inference</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<p>Use a proxy <span class="math inline">\(q_\psi\)</span> of <span class="math inline">\(p_\theta(\mathbf{Z}\,|\,\mathbf{Y})\)</span> minimizing a divergence in a class <span class="math inline">\(\mathcal{Q}\)</span> (e.g, KÃ¼llback-Leibler)</p>
<p><span class="math display">\[\begin{equation*}
q_\psi(\mathbf{Z})^\star \in \arg\min_{q\in\mathcal{Q}} KL\left(q(\mathbf{Z}), p_{\theta}(\mathbf{Z} | \mathbf{Y})\right).
\end{equation*}\]</span></p>
<p>and maximize the ELBO (Evidence Lower BOund)</p>
<p><span class="math display">\[\begin{equation*}
J(\theta, \psi) = \mathbb{E}_{\psi} [\log p_\theta(\mathbf{Y}, \mathbf{Z})] + \mathcal{H}[q_\psi(\mathbf{Z})]
\end{equation*}\]</span></p>
</section>
<section id="optimisation-methods" class="slide level2">
<h2>Optimisation methods</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<ul>
<li><p>Standard approach: alternative maximization of the ELBO: VE-step and M-step.</p></li>
<li><p>Brute-force approach:</p>
<p><span class="math display">\[(\theta^{(t+1)}, \psi^{(t+1)}) = (\theta^{(t+1)}, \psi^{(t+1)}) + \eta_t \nabla_{\theta, \psi} J(\theta^{(t)}, \psi^{(t)})\]</span></p></li>
</ul>
</section>
<section id="estimator-properties" class="slide level2">
<h2>Estimator properties</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<ul>
<li><p>Variance is available for the regression coefficient of PLN model (only) <span class="math inline">\(\implies\)</span> <strong>confidence intervals</strong> and <strong>hypothesis testing</strong>.</p></li>
<li><p>Estimator may be biased for some models</p></li>
</ul>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="textttpyplnmodels" class="title-slide slide level1 center">
<h1><span class="math inline">\(\texttt{pyPLNmodels}\)</span></h1>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
</section>
<section id="about-plnmodels" class="slide level2">
<h2>About PLNmodels</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div>
<ul>
<li class="fragment"><span class="math inline">\(\texttt{pyPLNmodels}\)</span> is the equivalent python package of the R package <span class="math inline">\(\texttt{PLNmodels}\)</span> <span class="citation" data-cites="PLNmodels">(<a href="#/references" role="doc-biblioref" onclick="">Chiquet, Mariadassou, et Robin 2021</a>)</span>.</li>
<li class="fragment">With some hindsight</li>
<li class="fragment">Emphasis is given on the optimization: simple and efficient.</li>
<li class="fragment">Implementation of <span class="math inline">\(\texttt{PlnAR}\)</span> and <span class="math inline">\(\texttt{ZIPlnPCA}\)</span> models</li>
</ul>
</div>
</section>
<section id="how-to-fit-a-model" class="slide level2">
<h2>How to fit a model</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<h4 id="like-r-formula">Like R-formula</h4>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1"></a><span class="im">from</span> pyPLNmodels <span class="im">import</span> Pln</span>
<span id="cb10-2"><a href="#cb10-2"></a>pln <span class="op">=</span> Pln.from_formula(<span class="st">"endog ~ labels "</span>, data <span class="op">=</span> rna)</span>
<span id="cb10-3"><a href="#cb10-3"></a>_ <span class="op">=</span> pln.fit()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Setting the offsets to zero.
Fitting a Pln model with full covariance.
Intializing parameters ...
Initialization finished.
Maximum number of iterations (400)  reached in 2.5 seconds.
Last  criterion = 5.4e-06 . Required tolerance = 1e-06</code></pre>
</div>
</div>
<div class="fragment">
<!-- -->
<h4 id="explicitly">Explicitly</h4>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1"></a><span class="im">from</span> pyPLNmodels <span class="im">import</span> Pln</span>
<span id="cb12-2"><a href="#cb12-2"></a>pln <span class="op">=</span> Pln(rna[<span class="st">"endog"</span>], exog <span class="op">=</span> rna[<span class="st">"labels_1hot"</span>], add_const <span class="op">=</span> <span class="va">False</span>)</span>
<span id="cb12-3"><a href="#cb12-3"></a>_ <span class="op">=</span> pln.fit()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Setting the offsets to zero.
Fitting a Pln model with full covariance.
Intializing parameters ...
Initialization finished.
Maximum number of iterations (400)  reached in 2.4 seconds.
Last  criterion = 5.4e-06 . Required tolerance = 1e-06</code></pre>
</div>
</div>
</div>
</section>
<section class="slide level2">

<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<h3 id="attributes-and-method">Attributes and method</h3>
<p>Available when calling <span class="math inline">\(\texttt{print(model)}\)</span>`:</p>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1"></a><span class="bu">print</span>(pln)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A multivariate Pln with full covariance.
======================================================================
     Loglike   Dimension    Nb param         BIC         AIC
   -226568.5         100        5350 245046.7454    231918.5

======================================================================
* Useful attributes
    .latent_variables .latent_positions .model_parameters .latent_parameters .optim_details
* Useful methods
    .transform() .show() .predict() .sigma() .projected_latent_variables() .plot_correlation_circle() .biplot() .viz() .pca_pairplot() .plot_expected_vs_true()
* Additional attributes for Pln are:
    None
* Additional methods for Pln are:
    .get_variance_coef() .get_confidence_interval_coef() .summary() .get_coef_p_values() .plot_regression_forest()</code></pre>
</div>
</div>
</section>
<section id="insights" class="slide level2">
<h2>Insights</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1"></a>pln.show(figsize <span class="op">=</span> (<span class="dv">10</span>, <span class="dv">8</span>))</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-14-output-1.png" width="817" height="680"></p>
</div>
</div>
</section>
<section id="insights-1" class="slide level2">
<h2>Insights</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1"></a>pln.plot_regression_forest(figsize <span class="op">=</span> (<span class="dv">11</span>,<span class="dv">6</span>))</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-15-output-1.png" width="1054" height="574"></p>
</div>
</div>
</section>
<section class="slide level2">

<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">Genes correlation</a></li><li><a href="#tabset-2-2">Network visualization</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<div>
<ul>
<li class="fragment"><p>The <span class="math inline">\(\texttt{PlnNetwork}\)</span> allows to infer genes that are highly correlated, by imposing a constraint:</p></li>
<li class="fragment"><p><span class="math display">\[\| \Sigma^{-1} \|_1 \leq \text{penalty}\]</span></p></li>
<li class="fragment"><p><span class="math display">\[\Sigma^{-1}_{ij} = \text{Corr}(Z_i, Z_j)\]</span></p></li>
<li class="fragment"><p>Still differentiable ELBO.</p></li>
</ul>
</div>
</div>
<div id="tabset-2-2">
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1"></a><span class="im">from</span> pyPLNmodels <span class="im">import</span> PlnNetwork</span>
<span id="cb18-2"><a href="#cb18-2"></a>net <span class="op">=</span> PlnNetwork(rna[<span class="st">"endog"</span>], penalty <span class="op">=</span> <span class="dv">2000</span>, compute_offsets_method <span class="op">=</span> <span class="st">"logsum"</span>).fit()</span>
<span id="cb18-3"><a href="#cb18-3"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">1</span>, figsize <span class="op">=</span> (<span class="dv">12</span>, <span class="dv">5</span>))</span>
<span id="cb18-4"><a href="#cb18-4"></a>net.viz_network(ax<span class="op">=</span>ax)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

</div>
</div>
</div>
</section>
<section id="" class="slide level2 output-location-slide"><div class="cell output-location-slide" data-execution_count="16">
<div class="cell-output cell-output-stdout">
<pre><code>Setting the offsets as the log of the sum of `endog`.
Fitting a PlnNetwork model with with penalty 2000
Intializing parameters ...
Initialization finished.
Maximum number of iterations (400)  reached in 2.8 seconds.
Last  criterion = 5.34e-06 . Required tolerance = 1e-06</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-16-output-2.png" width="912" height="409"></p>
</div>
</div></section><section id="predict-newly-observed-data" class="slide level2">
<h2>Predict newly observed data</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<p>The <span class="math inline">\(\texttt{PlnLDA}\)</span> allows to predict newly observed data. It performs LDA on the latent space.</p>
</section>
<section id="predict-newly-observed-data-1" class="slide level2">
<h2>Predict newly observed data</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="panel-tabset">
<ul id="tabset-3" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-3-1">Training</a></li><li><a href="#tabset-3-2">Confusion matrix</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1">
<div class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb20" data-code-line-numbers="1-4|5-6"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1"></a><span class="im">from</span> pyPLNmodels <span class="im">import</span> PlnLDA, plot_confusion_matrix</span>
<span id="cb20-2"><a href="#cb20-2"></a>ntrain <span class="op">=</span> <span class="dv">500</span></span>
<span id="cb20-3"><a href="#cb20-3"></a>endog_train, endog_test <span class="op">=</span> rna[<span class="st">"endog"</span>][:ntrain],rna[<span class="st">"endog"</span>][ntrain:]</span>
<span id="cb20-4"><a href="#cb20-4"></a>labels_train, labels_test <span class="op">=</span> rna[<span class="st">"labels"</span>][:ntrain], rna[<span class="st">"labels"</span>][ntrain:]</span>
<span id="cb20-5"><a href="#cb20-5"></a>lda <span class="op">=</span> PlnLDA(endog_train, clusters <span class="op">=</span> labels_train).fit()</span>
<span id="cb20-6"><a href="#cb20-6"></a>pred_test <span class="op">=</span> lda.predict_clusters(endog_test)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Setting the offsets to zero.
Fitting a PlnLDA model with full covariance.
Intializing parameters ...
Initialization finished.
Maximum number of iterations (400)  reached in 2.2 seconds.
Last  criterion = 5.4e-06 . Required tolerance = 1e-06
Setting the offsets to zero.
Doing a VE step.
Done!
Setting the offsets to zero.
Doing a VE step.
Done!
Setting the offsets to zero.
Doing a VE step.
Done!</code></pre>
</div>
</div>
</div>
<div id="tabset-3-2">
<div class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1"></a>plot_confusion_matrix(pred_test, labels_test)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-20-output-1.png" width="507" height="490"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section class="slide level2">

<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div class="panel-tabset">
<ul id="tabset-4" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-4-1">Zero-inflation investigation</a></li><li><a href="#tabset-4-2">Output</a></li></ul>
<div class="tab-content">
<div id="tabset-4-1">
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="im">from</span> pyPLNmodels <span class="im">import</span> ZIPln</span>
<span id="cb23-3"><a href="#cb23-3"></a>zi <span class="op">=</span> ZIPln(rna[<span class="st">"endog"</span>], exog <span class="op">=</span> rna[<span class="st">"labels_1hot"</span>], add_const <span class="op">=</span> <span class="va">False</span>).fit()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Setting the offsets to zero.
Fitting a ZIPln model with full covariance.
Intializing parameters ...
Initialization finished.
Maximum number of iterations (400)  reached in 5.4 seconds.
Last  criterion = 5.4e-06 . Required tolerance = 1e-06</code></pre>
</div>
</div>
</div>
<div id="tabset-4-2">
<div class="cell" data-execution_count="22">
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-22-output-1.png" width="873" height="643"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="zero-inflated-pln-textttzipln" class="title-slide slide level1 center">
<h1>Zero Inflated PLN: <span class="math inline">\(\texttt{ZIPln}\)</span></h1>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
</section>
<section id="zero-inflated-pln" class="slide level2">
<h2>Zero Inflated PLN</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div>
<ul>
<li class="fragment"><p>Use two latent vectors <span class="math inline">\(\mathbf{W}_i\)</span> and <span class="math inline">\(\mathbf{Z}_i\)</span> to model excess of zeroes and dependence structure <span class="math display">\[\begin{array}{rrl}
\text{PLN latent space} &amp;  \boldsymbol{Z}_i  = (Z_{ij})_{j=1\dots p} &amp; \sim  \mathcal{N}(\mathbf{x}_i^\top \mathbf{B}, \mathbf{\Sigma}), \\[1.5ex]
\text{excess of zeros} &amp;  \boldsymbol{W}_{i} = (W_{ij})_{j=1\dots p} &amp; \sim \otimes_{j=1}^p \cal B(\pi_{ij}),   \\[1.5ex]
\text{observation space} &amp;  Y_{ij} \, | \, W_{ij}, Z_{ij} &amp; \sim (1-W_{ij})\mathcal{P} \left(\exp\{o_{ij} + Z_{ij}\}\right),\\
\end{array}\]</span></p></li>
<li class="fragment"><p><span class="math inline">\(\rightsquigarrow\)</span> Better handling of zeros +additional interpretable parameters.</p></li>
<li class="fragment"><p>Still overdispersion <span class="math display">\[\mathbb V[Y_{ij}] &gt; \mathbb E[Y_{ij}]\]</span></p></li>
</ul>
</div>
</section>
<section id="zero-inflation-modelling" class="slide level2">
<h2>Zero-inflation modelling</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<p><span class="math display">\[  \pi_{ij}  = \sigma( \boldsymbol X^{0}\boldsymbol B^0)_{ij}, ~ \boldsymbol X^0 \in \mathbb R^{n \times d_0}, ~ \boldsymbol B^0 \in \mathbb R^{d_0\times p}\]</span></p>
<div>
<ul>
<li class="fragment">Other modelling possibilities (one for all, one per dimension, etc. )</li>
</ul>
</div>
</section>
<section id="identifiabiliy" class="slide level2">
<h2>Identifiabiliy</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<p>Identifiability of the model is ensured if the covariates <span class="math inline">\(X^0\)</span> are full rank.</p>
</section>
<section id="distribution-of-w_ij-y_ij" class="slide level2">
<h2>Distribution of <span class="math inline">\(W_{ij}| Y_{ij}\)</span></h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<p><span class="math display">\[W_{ij} | Y_{ij}  \sim \mathcal{B}\left(\frac{\pi_{ij}}{ \varphi\left(\mathbf{x}_i^\top \boldsymbol B_j, \Sigma_{jj}\right)
\left(1 - \pi_{ij}\right) + \pi_{ij}}\right) \boldsymbol 1_{\{Y_{ij} = 0\}}\]</span></p>
<p>with <span class="math inline">\(\varphi(\mu,\sigma^2) = \mathbb E \left[ \exp(-X)\right], ~ X \sim \mathcal L \mathcal N \left( \mu, \sigma^2\right)\)</span>.</p>
</section>
<section id="motivation-of-textttzipln-microcosm-dataset-microcosm" class="slide level2">
<h2>Motivation of <span class="math inline">\(\texttt{ZIPln}\)</span>: microcosm dataset <span class="citation" data-cites="microcosm">(<a href="#/references" role="doc-biblioref" onclick="">Mariadassou et al. 2023</a>)</span></h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
</section>
<section class="slide level2">

<div class="footer">

</div>
<ul>
<li>Data: sequencing of the microbiota of 45 dairy cows</li>
</ul>
<div class="column fragment" style="width:50%;">
<ul>
<li>from 4 sites:
<ul>
<li>vagina</li>
<li>mouth</li>
<li>nose</li>
<li>milk</li>
</ul></li>
</ul>
</div>
<div class="column fragment" style="width:50%;">
<ul>
<li>at 4 time points:
<ul>
<li>1 week before calving</li>
<li>1 month after</li>
<li>3 months after</li>
<li>7 months after</li>
</ul></li>
</ul>
</div>
</section>
<section id="data-description-4" class="slide level2">
<h2>Data description</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<ul>
<li>After data preprocessing <span class="math inline">\(\implies p = 1209\)</span> Amplicon Sequence Variants and <span class="math inline">\(96\)</span> % zeros.</li>
</ul>
<div class="cell" data-execution_count="24">
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-24-output-1.png" width="923" height="532"></p>
</div>
</div>
</section>
<section id="textttpln-versus-textttzipln" class="slide level2">
<h2><span class="math inline">\(\texttt{Pln}\)</span> versus <span class="math inline">\(\texttt{ZIPln}\)</span></h2>
<div class="footer">

</div>
<div class="panel-tabset">
<ul id="tabset-5" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-5-1">Fitting</a></li><li><a href="#tabset-5-2">Latent space</a></li><li><a href="#tabset-5-3"><span class="math inline">\(\Sigma\)</span></a></li></ul>
<div class="tab-content">
<div id="tabset-5-1">
<div class="cell" data-results="hide" data-execution_count="26">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1"></a>zi <span class="op">=</span> ZIPln.from_formula(<span class="st">"endog ~ site + time"</span>, micro).fit()</span>
<span id="cb25-2"><a href="#cb25-2"></a>pln <span class="op">=</span> Pln.from_formula(<span class="st">"endog ~ site + time"</span>, micro).fit()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>`exog_inflation` and `exog` are set to the same array. If you need different `exog_inflation`, specify it with a pipe: '|' like in the following: endog ~ 1 + x | x + y 
Setting the offsets to zero.
Now dataset of size torch.Size([300, 492]).
Fitting a ZIPln model with full covariance.
Intializing parameters ...
Initialization finished.
Maximum number of iterations (400)  reached in 9.7 seconds.
Last  criterion = 5.48e-06 . Required tolerance = 1e-06
Setting the offsets to zero.
Now dataset of size torch.Size([300, 492]).
Fitting a Pln model with full covariance.
Intializing parameters ...
Initialization finished.
Maximum number of iterations (400)  reached in 5.4 seconds.
Last  criterion = 5.37e-06 . Required tolerance = 1e-06</code></pre>
</div>
</div>
</div>
<div id="tabset-5-2">
<div class="cell" data-execution_count="27">
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-27-output-1.png" width="841" height="617"></p>
</div>
</div>
</div>
<div id="tabset-5-3">
<div class="cell" data-execution_count="28">
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-28-output-1.png" width="949" height="605"></p>
</div>
</div>
</div>
</div>
</div>
</section></section>
<section id="autoregressive-pln-textttplnar" class="title-slide slide level1 center">
<h1>Autoregressive Pln: <span class="math inline">\(\texttt{PlnAR}\)</span></h1>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div>
<ul>
<li class="fragment">Assume dependency between individuals <span class="math inline">\(\implies\)</span> Time series or spatial 1D: <span class="math display">\[
  \begin{aligned}
  \mathbf{Z}_{i+1}|\mathbf Z_i &amp; \sim \Phi \mathbf Z_i  + \epsilon_i \\
  \epsilon_i &amp; \sim \mathcal{N}(\Phi \mu_i, \boldsymbol\Sigma_{\epsilon} )\\
  \mathbf{Y}_{i+1} &amp; \sim \mathcal{P}(\exp(\mathbf Z_{i+1}))
  \end{aligned}
  \]</span></li>
<li class="fragment">Constraint: <span class="math display">\[ \boldsymbol \Sigma_{\epsilon} = \Sigma - \Phi \boldsymbol \Sigma \Phi^\top \succ 0\]</span></li>
<li class="fragment">Marginal laws: <span class="math display">\[\mathbf{Z}_i \sim \mathcal{N}(\mu_i, \boldsymbol\Sigma)\]</span></li>
</ul>
</div>
</section>

<section>
<section id="autoregressive-constraints" class="title-slide slide level1 center">
<h1>Autoregressive constraints</h1>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<p><span class="math display">\[\boldsymbol \Sigma_{\epsilon} = \boldsymbol \Sigma - \Phi\boldsymbol  \Sigma \Phi^\top \succ 0\]</span></p>
<div>
<ul>
<li class="fragment"><p>âEasyâ cases:</p>
<ul>
<li class="fragment"><span class="math inline">\(\boldsymbol \Sigma\)</span> diagonal and <span class="math inline">\(1 \succ \Phi \succ 0\)</span> diagonal</li>
<li class="fragment"><span class="math inline">\(\boldsymbol \Sigma\)</span> full and <span class="math inline">\(1 &gt; \Phi &gt; 0\)</span> scalar</li>
</ul></li>
<li class="fragment"><p>Non positive definite case:</p>
<ul>
<li class="fragment"><span class="math inline">\(\boldsymbol \Sigma\)</span> full and <span class="math inline">\(1 \succ \Phi \succ 0\)</span> diagonal <span class="math inline">\(\implies\)</span> <span class="math inline">\(\boldsymbol \Sigma_{\epsilon}\)</span> non definite positive</li>
</ul></li>
<li class="fragment"><p>More complex case:</p>
<ul>
<li class="fragment"><span class="math inline">\(\boldsymbol \Sigma\)</span> full and <span class="math inline">\(1 \succ \Phi \succ 0\)</span> full</li>
</ul></li>
</ul>
</div>
</section>
<section id="autoregressive-constraints-1" class="slide level2">
<h2>Autoregressive constraints</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<p><span class="math display">\[
\begin{array}{c|cc}
\Phi \ \backslash \ \Sigma  &amp; \text{Diagonal} &amp; \text{Full} \\
\hline
\text{Scalar}    &amp; \checkmark &amp; \checkmark \\
\text{Diagonal}  &amp; \checkmark &amp; \times \\
\text{Full}      &amp; \times &amp; ?? \\
\end{array}
\]</span></p>
</section>
<section id="motivation-crossover-dataset" class="slide level2">
<h2>Motivation: crossover dataset</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div>
<ul>
<li class="fragment"><p>Data: number of exchange of genetic material between two DNA strand in a loci.</p></li>
<li class="fragment"><p>All the 27 chromosomes placed end to end <span class="math inline">\(\implies\)</span> <strong>500 loci</strong>, for <span class="math inline">\(\textbf{4 different species of sheep}\)</span>.</p></li>
<li class="fragment"><p><span class="math inline">\(\implies\)</span> 4 time series of 500 loci. Each loci depends on the previous one.</p></li>
</ul>
</div>
</section>
<section id="textttplnar-diagonal" class="slide level2">
<h2><span class="math inline">\(\texttt{PlnAR}\)</span> diagonal</h2>
<div class="footer">

</div>
<div class="panel-tabset">
<ul id="tabset-6" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-6-1">Fitting</a></li><li><a href="#tabset-6-2">Latent space</a></li><li><a href="#tabset-6-3">Autoregressive structure</a></li></ul>
<div class="tab-content">
<div id="tabset-6-1">
<div class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1"></a><span class="im">from</span> pyPLNmodels <span class="im">import</span> load_crossover, PlnAR</span>
<span id="cb27-2"><a href="#cb27-2"></a>cross <span class="op">=</span> load_crossover()</span>
<span id="cb27-3"><a href="#cb27-3"></a>ar <span class="op">=</span> PlnAR(cross[<span class="st">"endog"</span>], ar_type <span class="op">=</span> <span class="st">"diagonal"</span>).fit()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Returning crossover dataset of size (500, 4)
Setting the offsets to zero.
Fitting a PlnAR model with autoregressive type diagonal.
Intializing parameters ...
Initialization finished.
Maximum number of iterations (400)  reached in 1.2 seconds.
Last  criterion = 5.4e-06 . Required tolerance = 1e-06</code></pre>
</div>
</div>
</div>
<div id="tabset-6-2">
<div class="cell" data-execution_count="30">
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-30-output-1.png" width="781" height="579"></p>
</div>
</div>
</div>
<div id="tabset-6-3">
<div class="cell" data-execution_count="31">
<div class="cell-output cell-output-display">
<p><img data-src="slides_files/figure-revealjs/cell-31-output-1.png" width="739" height="631"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="conclusion" class="slide level2">
<h2>Conclusion</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div>
<ul>
<li class="fragment"><span class="math inline">\(\texttt{pyPLNmodels}\)</span> easy to use for many statistical tasks for count data</li>
<li class="fragment">Focus on optimization and scalability when possible.</li>
<li class="fragment">Majority of models gives out of memory error for <span class="math inline">\(n,p&gt; 3000\)</span> in CPU.</li>
</ul>
</div>
</section>
<section id="whats-next" class="slide level2">
<h2>Whatâs next?</h2>
<div class="footer">
<span class="math inline">\(\texttt{pip install pyPLNmodels}\)</span>
</div>
<div>
<ul>
<li class="fragment">Implement a <span class="math inline">\(\texttt{PlnAR}\)</span> with full autoregressive structure.</li>
<li class="fragment">Implement a collection of <span class="math inline">\(\texttt{PlnNetwork}\)</span> and <span class="math inline">\(\texttt{PlnMixture}\)</span> to directly find the BIC of the best model.</li>
<li class="fragment">Make it worldwide</li>
</ul>
</div>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>
<p><img src="_extensions/davidcarayon/inrae/ressources/assets/bloc-etat.png" class="slide-logo"></p>
<div class="footer footer-default">
<p>Pied de page</p>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-AiH89" class="csl-entry" role="listitem">
Aitchison, J., et C. H. Ho. 1989. <span>Â«&nbsp;The multivariate Poisson-log normal distribution&nbsp;Â»</span>. <em>Biometrika</em> 76 (4): 643â53.
</div>
<div id="ref-PLNPCA" class="csl-entry" role="listitem">
Chiquet, Julien, Mahendra Mariadassou, et StÃ©phane Robin. 2018. <span>Â«&nbsp;Variational inference for probabilistic Poisson PCA&nbsp;Â»</span>. <em>The Annals of Applied Statistics</em> 12: 2674â98. <a href="http://dx.doi.org/10.1214/18-AOAS1177">http://dx.doi.org/10.1214/18-AOAS1177</a>.
</div>
<div id="ref-PLNnetwork" class="csl-entry" role="listitem">
âââ. 2019. <span>Â«&nbsp;Variational inference for sparse network reconstruction from count data&nbsp;Â»</span>. In <em>Proceedings of the 19th International Conference on Machine Learning (ICML 2019)</em>.
</div>
<div id="ref-PLNmodels" class="csl-entry" role="listitem">
âââ. 2021. <span>Â«&nbsp;The Poisson-Lognormal Model as a Versatile Framework for the Joint Analysis of Species Abundances&nbsp;Â»</span>. <em>Frontiers in Ecology and Evolution</em> 9. <a href="https://doi.org/10.3389/fevo.2021.588292">https://doi.org/10.3389/fevo.2021.588292</a>.
</div>
<div id="ref-microcosm" class="csl-entry" role="listitem">
Mariadassou, Mahendra, Laurent-Xavier Nouvel, Fabienne Constant, Diego P. Morgavi, Lucie Rault, Sarah Barbey, Emmanuelle Helloin, et al. 2023. <span>Â«&nbsp;<span>Microbiota members from body sites of dairy cows are largely shared within individual hosts throughout lactation but sharing is limited in the herd</span>&nbsp;Â»</span>. <em><span>Animal Microbiome</span></em> 5 (1): 32. <a href="https://doi.org/10.1186/s42523-023-00252-w">https://doi.org/10.1186/s42523-023-00252-w</a>.
</div>
<div id="ref-scMARK" class="csl-entry" role="listitem">
Swechha, Dylan Mendonca, Octavian Focsa, J. Javier DÄ±Ìaz-MejÄ±Ìa, et Samuel Cooper. 2021. <span>Â«&nbsp;scMARK an <span>â</span>MNIST<span>â</span> like benchmark to evaluate and optimize models for unifying scRNA data&nbsp;Â»</span>. <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2021.12.08.471773">https://doi.org/10.1101/2021.12.08.471773</a>.
</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="slides_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="slides_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="slides_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="slides_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="slides_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="slides_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="slides_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="slides_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'fade',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 780,

        // Factor of the display size that should remain empty around the content
        margin: 0.2,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>

    <script type="text/javascript" charset="utf-8">
      $(document).ready(function() {
        // Add `data-footer` on title slide
        function addCustomFooter() {
          let titleSlide = $("section#title-slide");
          let titleSlideFooter = titleSlide.attr('data-footer');
          let footer = $('div.footer p');
          let globalFooterText = footer.html();

          if (titleSlide.hasClass('present')) {
            footer.html(titleSlideFooter);
          }

          Reveal.on('slidechanged', function(event) {
            if (event.currentSlide.matches('#title-slide')) {
              footer.html(titleSlideFooter);
            } else {
              footer.html(globalFooterText);
            }
          });
        }

        addCustomFooter();

        // Hide slide number on title slide
        function removeSlideNumber() {
          let slideNo = $('div.slide-number');
          slideNo.addClass('hide');

          Reveal.on('slidechanged', function(event) {
            if (Reveal.isFirstSlide()) {
              slideNo.addClass('hide');
            } else {
              slideNo.removeClass('hide');
            }
          });
        }

        removeSlideNumber();
      });
    </script>

    <style>
      .hide {
        display: none !important;
      }
    </style>
    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "CopiÃ©");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "CopiÃ©");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>